File: routes/startup_container.go
package routes

import (
	"fmt"
	"os"
	"os/exec"
	"server/method"
	"strings"

	"github.com/gofiber/fiber/v2"
)

func StartupContainer(app *fiber.App) {

	app.Get("/api/instance/startup/:bot_id/:admin_key/", func(c *fiber.Ctx) error {

		var bot_id = c.Params("bot_id")
		var admin_key = c.Params("admin_key")

		if !method.ValidateAdminKey(admin_key) {
			fmt.Println("[Startup] Erreur admin_key", admin_key, " n'est pas valide!")
			return c.Status(fiber.StatusBadRequest).SendString("Invalid admin_key!")
		}

		if _, err := os.Stat(method.PathResolve(method.ProcessCWD(), "ownihrz", bot_id)); os.IsNotExist(err) {
			fmt.Println("[Startup] Erreur bot_id", bot_id, " n'existe pas!")
			return c.Status(fiber.StatusBadRequest).SendString("Invalid bot_id!")
		}

		cliArray := []struct {
			L   string
			CWD string
		}{
			{
				L:   "rm -r dist",
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", bot_id),
			},

			{
				L:   "git pull",
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", bot_id),
			},

			{
				L:   "npx tsc",
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", bot_id),
			},

			{
				L:   strings.Replace("mv dist/index.js dist/{Code}.js", "{Code}", bot_id, 1),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", bot_id),
			},

			{
				L:   strings.Replace("pm2 start dist/{Code}.js -f", "{Code}", bot_id, 1),
				CWD: method.PathResolve(method.ProcessCWD(), "ownihrz", bot_id),
			},
		}

		for _, index := range cliArray {
			cmd := exec.Command("sh", "-c", index.L)
			cmd.Dir = index.CWD
			cmd.Stdout = os.Stdout
			cmd.Stderr = os.Stderr
			if err := cmd.Run(); err != nil {
				fmt.Print(err)
				return c.Status(fiber.StatusInternalServerError).SendString(err.Error())
			}
		}

		return c.SendStatus(200)
	})

}

File: src/Routes/v1/startup_container.ts
import { validateAdminKey } from '../../method/validateData.js';
import isContainerOn from '../../method/getContainerStatus.js';
import config from '../../method/getConfigData.js';

import { Request, Response } from 'express';
import { execSync } from "child_process";
import path from "node:path";
import fs from "node:fs";

export default {
    type: 'get',
    apiPath: '/api/v1/instance/startup/:bot_id/:admin_key/',
    run: async (req: Request, res: Response) => {

        const botId = req.params["bot_id"];
        const adminKey = req.params["admin_key"];

        if (!config?.api.apiToken) {
            console.log("Error: Failed to load config");
            return res.status(500).send("Failed to load config");
        }

        if (!validateAdminKey(adminKey)) {
            console.log("[Delete] Erreur admin_key n'est pas valide!");
            return res.status(403).send("Invalid admin_key!");
        };

        if (!fs.existsSync(path.join(process.cwd(), 'ownihrz', botId))) {
            console.log("[Delete] Erreur bot_id n'existe pas!");
            return res.status(403).send("Invalid bot_id!");
        };

        if (!await isContainerOn(botId)) {
            [
                {
                    line: 'rm -r -f dist',
                    cwd: path.join(process.cwd(), 'ownihrz', botId)
                },
                {
                    line: 'git pull',
                    cwd: path.join(process.cwd(), 'ownihrz', botId)
                },
                {
                    line: "bun install",
                    cwd: path.join(process.cwd(), 'ownihrz', botId),
                },
                {
                    line: `npx tsc`,
                    cwd: path.join(process.cwd(), 'ownihrz', botId)
                },
                {
                    line: `mv dist/index.js dist/${botId}.js`,
                    cwd: path.join(process.cwd(), 'ownihrz', botId)
                },
                {
                    line: `pm2 start dist/${botId}.js -f`,
                    cwd: path.join(process.cwd(), 'ownihrz', botId)
                },
            ].forEach((index) => {
                try {
                    execSync(index.line, { stdio: [0, 1, 2], cwd: index.cwd });
                } catch (e) {
                    console.log((e as string).split('\n'));
                }
            });
        } else {
            console.log('[Startup] Erreur tentative doublon!');
        }

        return res.sendStatus(200);
    },
};